### PR TITLE
refactor: descriptive request handling in dynamic configuration

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This changes the cluster configuration to be more descriptive on request handling. Where we previously only had a simple
 set of active partitions, we can now differentiate between different cases such as:
 - All partitions from 1..=partitionCount handle requests
 - All partitions from 1..=partitionCount handle requests, with some additional active partitions and some inactive ones.

 This makes it easier to keep track of the routing state during scale up (and later scale down). In particular, this
 allows idempotent scale up operations because we can now update the routing state immediately to reflect new
 partitions, something that wasn't possible when we only listed active partitions.

BREAKING CHANGE: This breaks compatibility, but only if the experimental feature flag for partition scaling is enabled. This is allowed for now.

Relates to https://github.com/camunda/camunda/issues/23215

### CODE REVIEW

**Review by npepinpe — COMMENTED** [review]
:melting_face: That was a rough onboarding - I'm really behind on the topology/dynamic config stuff, and the new partition scaling :smile: I don't mean the PR is difficult to understand, just everything led to more questions (but about existing stuff :upside_down_face:)

PR looks good, as usual code is easy to read and all :)

:question: I have one main question though, and kind of fundamental :sweat_smile: I don't really understand the purpose of the change. Let's say we just kept the following structure, where we showed always the current active and inactive partitions for routing, e.g.

```json
{
  "active": [1, 2, 3],
  "inactive": []
}

{
  "active": [1, 2, 3],
  "inactive": [4]
}

{
  "active": [1, 2, 3, 4],
  "inactive": []
}
```

Why is this less understandable than having a strategy and its fields (which need to be interpreted)? Is it for humans? For machine readability?

I haven't been so much in this topic, so maybe I'm just not used to how we monitor changes. Is the expectation that I call the cluster endpoint for the latest topology, and I can then understand from the strategy where in the scaling operation I am? So the expected results from successive calls would be:

```json
"requestHandling": {
  "strategy": "AllPartitions",
  "partitionCount": 1
}


"requestHandling": {
  "strategy": "Mixed",
  "base": {
    "strategy": "AllPartitions",
    "partitionCount": 1
  },
  "additional": {
    "strategy": "ActivePartitions",
    "activePartitions: [],
    "inactivePartitions": [2]
  }
}


"requestHandling": {
  "strategy": "AllPartitions",
  "partitionCount": 2
}
```

Or? I'm not really sure, to be honest. When would I get just `ActivePartitions`, like some of the tests check?

**Comment by lenaschoenburg** [comment]
> Why is this less understandable than having a strategy and its fields (which need to be interpreted)? Is it for humans? For machine readability?

A bit of both actually. We could have stayed with a simple list of partition ids but I saw some mentions of >100 partition deployments which is pushing the limits of what I'd store in a simple list that is gossipped around and meant to be human-readable. To overcome this, I wanted two different representations, one efficient one when all partitions from 1 to N are active and one that is easy to work during scaling where we list the partition ids directly because active and inactive partitions are not contiguous.

 We could have gotten away with just these two representations (and maybe that would have been better?) but I though why not make them somewhat composable with the _mixed_ strategy that combines a contiguous range of active partitions with sets of inactive + active.

> When would I get just ActivePartitions, like some of the tests check?

In practice never, so maybe it would be better to simplify a little bit and have just two:
```
AllPartitions(int partitionCount)
Mixed(int basePartitionCount, Set<Integer> newActive, Set<Integer> pending)
```

File: dist/src/main/java/io/camunda/zeebe/shared/management/ClusterApiUtils.java
Comment (lenaschoenburg) [reply dist/src/main/java/io/camunda/zeebe/shared/management/ClusterApiUtils.java:193]: The function is imported directly here: https://github.com/camunda/camunda/pull/23777/files/a93d3d93478d221fb06a832c2c385d7f51c59909#diff-b50af5636f9871f0d0050597de0c64971ff3315bb4d647169994a74d5014f8c0R11

Comment (npepinpe) [reply dist/src/main/java/io/camunda/zeebe/shared/management/ClusterApiUtils.java:193]: Ah, I missed it, I was really surprised :smile:

**Comment by npepinpe** [comment]
Thanks for the explanation. I hadn't considered encoding 100s of partitions. Considering partitions are typically contiguous, and we scale up one at a time, would encoding them as ranges make more sense? e.g.

```json
{
  "active": "[1,3]",
  "inactive": "[4,99]"
}
```

Or could we scale asynchronously in different orders?

**Review by npepinpe — APPROVED** [review]
Pre-approving after discussion that we will merge the `ActivePartitions` and `Mixed` strategies

**Pull request merged**
